File: web/src/lib/crypto/aead.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 PRSN EX Inc.
// AES-256-GCM (A256GCM, RFC 7518 §5.3). Returns/consumes ciphertext ‖ tag (the
// trailing 16 bytes are the GCM tag), matching the on-disk envelope layout and
// WebCrypto's own ct‖tag convention. AAD binds a ciphertext to its semantic
// context (file_id, folder scope); threading it through is the caller's job.

import type { Bytes } from './bytes';
import { IntegrityError, InvalidInputError } from './errors';

function importGcmKey(key: Bytes, usage: KeyUsage): Promise<CryptoKey> {
  if (key.length !== 32) throw new InvalidInputError('aes-256-gcm key must be 32 bytes');
  return crypto.subtle.importKey('raw', key, { name: 'AES-GCM' }, false, [usage]);
}

/** Seal plaintext under key (256-bit) + nonce (96-bit) with AAD; returns ct ‖ tag.
 *  The caller MUST use a unique nonce per (key, message). */
export async function aesGcmSeal(
  key: Bytes,
  nonce: Bytes,
  plaintext: Bytes,
  aad: Bytes,
): Promise<Bytes> {
  const k = await importGcmKey(key, 'encrypt');
  const ct = await crypto.subtle.encrypt(
    { name: 'AES-GCM', iv: nonce, additionalData: aad, tagLength: 128 },
    k,
    plaintext,
  );
  return new Uint8Array(ct);
}

/** Open ciphertextAndTag (ct ‖ tag) under key + nonce + AAD. Throws
 *  {@link IntegrityError} on tag mismatch, wrong key, wrong AAD, or tampering —
 *  the caller MUST NOT distinguish these. */
export async function aesGcmOpen(
  key: Bytes,
  nonce: Bytes,
  ciphertextAndTag: Bytes,
  aad: Bytes,
): Promise<Bytes> {
  const k = await importGcmKey(key, 'decrypt');
  try {
    const pt = await crypto.subtle.decrypt(
      { name: 'AES-GCM', iv: nonce, additionalData: aad, tagLength: 128 },
      k,
      ciphertextAndTag,
    );
    return new Uint8Array(pt);
  } catch {
    throw new IntegrityError('aes-256-gcm authentication failed');
  }
}

File: crypto/src/aead.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 PRSN EX Inc.

//! AES-256-GCM (`A256GCM`, RFC 7518 §5.3 / NIST SP 800-38D).
//!
//! The authenticated-encryption primitive for file ciphertext (Envelope §4),
//! encrypted names (§7.3), and the human-side wrap blob (§8.3). AAD binds a
//! ciphertext to its semantic context (file_id, folder scope) so the storage
//! layer cannot substitute one blob for another — the binding is the caller's
//! responsibility; this module just threads the AAD through.

use aes_gcm::aead::{Aead, KeyInit, Payload};
use aes_gcm::{Aes256Gcm, Key, Nonce};

use crate::error::{CryptoError, Result};

/// Seal `plaintext` under `key` (256-bit) and `nonce` (96-bit) with the given
/// AAD. Returns `ciphertext || tag` (the trailing 16 bytes are the GCM tag),
/// matching the on-disk envelope layout (Envelope §4.1: `[IV][ct][tag]`).
///
/// The caller MUST use a unique nonce per (key, message); GCM nonce reuse is
/// catastrophic. This module does not generate nonces — that is the envelope
/// layer's job (fresh random per file, or HKDF-derived per multipart chunk).
pub fn seal(key: &[u8; 32], nonce: &[u8; 12], plaintext: &[u8], aad: &[u8]) -> Result<Vec<u8>> {
    let cipher = Aes256Gcm::new(Key::<Aes256Gcm>::from_slice(key));
    cipher
        .encrypt(
            Nonce::from_slice(nonce),
            Payload {
                msg: plaintext,
                aad,
            },
        )
        .map_err(|_| CryptoError::InvalidInput("aes-256-gcm seal"))
}

/// Open `ciphertext_and_tag` (`ciphertext || tag`) under `key`, `nonce`, and AAD.
/// Returns the plaintext, or [`CryptoError::Authentication`] on tag mismatch,
/// wrong key, wrong AAD, or tampering — the caller MUST NOT distinguish these.
pub fn open(
    key: &[u8; 32],
    nonce: &[u8; 12],
    ciphertext_and_tag: &[u8],
    aad: &[u8],
) -> Result<Vec<u8>> {
    let cipher = Aes256Gcm::new(Key::<Aes256Gcm>::from_slice(key));
    cipher
        .decrypt(
            Nonce::from_slice(nonce),
            Payload {
                msg: ciphertext_and_tag,
                aad,
            },
        )
        .map_err(|_| CryptoError::Authentication)
}
